pycytominer/cyto_utils/DeepProfiler_processing.py

```diff
@@ -0,0 +1,160 @@
+        self.index_df = pd.read_csv(index_file)
+
+    def build_filenames(self):
+        """
+        Single cell profile file names indicated by plate, well, and site information
+        """
+        self.filenames = self.index_df.apply(
+            self.build_filename_from_index, axis="columns"
+        )
+        self.filenames = [
+            pathlib.PurePath(f"{self.profile_dir}/{x}") for x in self.filenames
+        ]
+
+    def build_filename_from_index(self, row):
+        plate = row["Metadata_Plate"]
+        well = row["Metadata_Well"]
+        site = row["Metadata_Site"]
+
+        filename = f"{plate}_{well}_{site}{self.file_extension}"
+        return filename
+
+    def extract_filename_metadata(self, npz_file, delimiter="_"):
+        """
+        Format: plate_well_site.npz
+        """
+        base_file = os.path.basename(npz_file).strip(".npz").split(delimiter)
+        site = base_file[-1]
+        well = base_file[-2]
+        plate = delimiter.join(base_file[:-2])
+
+        return {"site": site, "well": well, "plate": plate}
+
+    def setup_aggregate(self):
+        if not hasattr(self, "filenames"):
+            self.build_filenames()
+
+        self.file_aggregate = {}
+        for filename in self.filenames:
+            file_info = self.extract_filename_metadata(filename, self.file_delimiter)
+            file_key = file_info[self.aggregate_on]
+
+            if self.aggregate_on == "site":
+                file_key = (
+                    f"{file_info['plate']}_{file_info['well']}_{file_info['site']}"
+                )
+
+            if self.aggregate_on == "well":
+                file_key = f"{file_info['plate']}_{file_info['well']}"
+
+            if file_key in self.file_aggregate:
+                self.file_aggregate[file_key]["files"].append(filename)
+            else:
+                self.file_aggregate[file_key] = {}
+                self.file_aggregate[file_key]["files"] = [filename]
+
+            self.file_aggregate[file_key]["metadata"] = file_info
+
+    def aggregate_deep(self):
+        if not hasattr(self, "file_aggregate"):
+            self.setup_aggregate()
+
+        self.aggregated_profiles = []
+        self.aggregate_merge_col = f"Metadata_{self.aggregate_on.capitalize()}_Position"
+
+        for metadata_level in self.file_aggregate:
+            df = pd.concat(
+                [load_npz(x) for x in self.file_aggregate[metadata_level]["files"]]
+            )
+            meta_df = pd.DataFrame(
+                self.file_aggregate[metadata_level]["metadata"], index=[0]
+            )
+            meta_df.columns = [f"Metadata_{x.capitalize()}" for x in meta_df.columns]
+
+            if self.aggregate_on == "well":
+                meta_df = meta_df.drop("Metadata_Site", axis="columns")
+
+            features = df.columns.tolist()
+            df = df.assign(Metadata_Aggregate_On=self.aggregate_on)
+            df = aggregate(
```

Comment:
This part is giving me Errors @gwaygenomics. 
The features list is a list of all columns (Metadata and profiles) this gives an error in aggregate because it expects only the profiles, I think. 
If you don't know the solution, I will work on it

Reply:
what is the error?

at this point, df should just be features, plus a column Metadata_Aggregate_On - but the features should not include this variable. (the metadata should all be in metadata_df)